### DISCUSSION TITLE
[resolved] Problem z dyskusjami REST API

### DISCUSSION THREAD

**Discussion by creativepl-php** [allegro/allegro-api#4332] [Suggestions - resolved]
Występuje problem: 
 - [X] z zasobem REST API
 - [X] ze środowiskiem produkcyjnym
 - [ ] ze środowiskiem testowym
 - [ ] z logowaniem lub autoryzacją
 - [ ] na portalu [https://developer.allegro.pl/](https://developer.allegro.pl/)
 - [ ] inny

:beetle: **Opis**
Cześć, pobierając dyskusje po API wszystko jest ok - wyświetlam je sobie u siebie na stronie, natomiast nie wiem kiedy dyskusja została "odświeżona" czyli albo ja albo kupujący udzielili się w dyskusji. Dyskusje zwracane są po sortowaniu wg. Daty rozpoczecia dyskusji, a nie od najświeższej wiadomości. Dodatkowo parametr "status" nie ulega zmianie gdy sprzedający lub kupujący odpowiedzą np gdy kupujący zgłosi powinien być status "NEW", gdy sprzedający odpowie to "SELLER_REPLY" a gdy kupujący odpowie to "BUYER_REPLY". Jest po prostu od początku trwania dyskusji do końca "ONGOING" do "CLOSED". Rodzi to pewne problemy. Nie jest również zwracana data ostatniej odpowiedzi np last_message jako datetime, żeby sprawdzać czy doszła nowa wiadomość. 

Oznacza to, że jakbym chciał sprawdzać czy w dyskusji pojawiło się coś nowego, to musiałbym 
1. Pobrać wszystkie dostępne dyskusje
2. Przy tych które nie mają statusu CLOSED loopowac wszystkie otwarte (przy założeniu że otwartych jest 200 dyskusji) to wysyłać 200 requestow o status dyskusji (/sale/disputes/ID) a później wysłać kolejny request o wiadomości (/sale/disputes/ID/messages)  i sprawdzać czy ostatnia data zgadza się z tą w mojej bazie - jeśli nie to wówczas oznaczałoby że doszła nowa wiadomość. 
3. Reasumując żeby wiedzieć o nowych wiadomościach w dyskusjach musiałbym wykorzystać CRON np co 5 minut i co 5 minut nawalać w was 400 requestami bo chce być na bieżąco z dyskusjami.

Czy macie na to jakieś rozwiązanie, czy mogę spamować was co 5 minut?

Kolejna sprawa, że nie wiem czy mój serwer timeouta nie wywali przy tylu requestach, czy PHP zdąży wszystko przemielić. Tak czy inaczej - API nie jest wydajne w tym zakresie.

**Comment by AureliuszBrussy** [comment]
Aktualnie nie ma innego sposobu na sprawdzenie nowych wiadomości, niż ten, który przedstawiasz.

Dziękuję za Twoje uwagi. Już wcześniej otrzymaliśmy podobne sugestie. Dopisuję do nich Twój głos.

**Comment by creativepl-php** [comment]
@AureliuszBrussy generalnie wymyśliłem coś innego - przy nowej dyskusji oraz nowej wiadomości w dyskusji wysyłacie wiadomość e-mail. Rozwiązaniem byłoby więc przechwytywanie maili po IMAP, tam jest login użytkownika w tytule maila więc można dojść który ticket ma nową odpowiedź (można też zresztą przeparsować HTML maila i wyciągnąć id dyskusji). W każdym razie no wypadałoby, żeby API miało to.

**Comment by AureliuszBrussy** [comment]
@damianszymanskiphp tak, jest to alternatywne rozwiązanie, jednak wykracza poza obsługę samego API Allegro. W takich przypadkach należy jednak mieć na uwadze, że np. jeżeli użytkownik ma włączoną daną dyskusję na www - nie wyślemy powiadomienia o pojawieniu się nowej wiadomości.

**Comment by creativepl-php** [comment]
@AureliuszBrussy "użytkownik" znaczy seller czy buyer? Bo jak buyer to proponuję zmienić, żeby wysyłało :)

**Comment by AureliuszBrussy** [comment]
I kupujący i sprzedający nie otrzymają powiadomienia, jeżeli mają wyświetloną Dyskusję i w takim momencie otrzymają wiadomość.

**Comment by pkrzemo** [comment]
@AureliuszBrussy i to jest właśnie beznadziejne...
Wystarczy mieć otwartych kilka zakładek, albo mieć "na wierzchu" inny program aby nie zauważyć nowej wiadomości od kupującego... A potem allegro rozlicza z czasu odpowiedzi w dyskusji... beznadzieja!
Jesteście takie biedaki że nie stać was na wysłanie głupiego e-maila ZA KAŻDYM RAZEM gdy KUPUJĄCY coś napisze w dyskusji?
Czy po prostu to taki rodzaj złośliwości względem sprzedającego?

**Comment by AureliuszBrussy** [comment]
@pkrzemo jeżeli masz uwagi co do funkcjonalności Allegro nie związanych z API - proszę, skorzystaj z [formularza kontaktowego](https://allegro.pl/pomoc/kontakt?kategoria=a0f7ce34-c6f6-43e6-af7d-ad39bacce619&subjectId=517305f2-5ea1-4b3b-9080-d4cc960ab6a1).

**Comment by AureliuszBrussy** [comment]
Wdrożyliśmy zmiany usprawniające proces odczytywania nowych wiadomości w dyskusjach. Więcej możesz przeczytać w [naszym newsie](https://github.com/allegro/allegro-api/issues/4846).
